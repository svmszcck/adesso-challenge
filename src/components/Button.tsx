import React from "react";
import { TouchableOpacity, Text, StyleSheet, ViewStyle } from "react-native";

import { ThemedText } from "@/components";
import { useThemeColor } from "@/hooks/useThemeColor";
import { FontSize } from "@/constants";

type ButtonProps = {
  title: string;
  accessibilityLabel?: string;
  onPress: () => void;
  style?: ViewStyle;
  isSmall?: boolean;
};

const CustomButton: React.FC<ButtonProps> = ({
  title,
  accessibilityLabel,
  onPress,
  style,
  isSmall,
}) => {
  const textColor = useThemeColor({}, "text");
  const backgroundColor = useThemeColor({}, "background");

  return (
    <TouchableOpacity
      onPress={onPress}
      style={[
        styles.buttonContainer,
        { backgroundColor: textColor, paddingVertical: isSmall ? 6 : 10 },
        style,
      ]}
      accessibilityLabel={accessibilityLabel || title}
      accessibilityRole="button"
      accessible
      activeOpacity={0.8}
    >
      <ThemedText style={[styles.buttonText, { color: backgroundColor }]}>
        {title}
      </ThemedText>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  buttonContainer: {
    elevation: 8,
    backgroundColor: "#009688",
    borderRadius: 10,
    paddingVertical: 10,
    paddingHorizontal: 12,
  },
  buttonText: {
    fontSize: FontSize.SMALL,
    fontWeight: "bold",
    alignSelf: "center",
  },
});

export default CustomButton;
