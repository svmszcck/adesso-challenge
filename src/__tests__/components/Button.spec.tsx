import React from "react";
import { render, screen, fireEvent } from "@testing-library/react-native";

import Button from "@/components/Button";

jest.mock("@/hooks/useThemeColor", () => ({
  useThemeColor: jest.fn((_, colorName) =>
    colorName === "text" ? "#ffffff" : "#009688"
  ),
}));

describe("Custom Button Component", () => {
  const mockOnPress = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly with given props", () => {
    render(<Button title="Click Me" onPress={mockOnPress} />);

    expect(screen.getByText("Click Me")).toBeTruthy();
  });

  it("renders correctly with accessibility label", () => {
    render(
      <Button
        title="Click Me"
        onPress={mockOnPress}
        accessibilityLabel="Click Button"
      />
    );

    expect(screen.getByLabelText("Click Button")).toBeTruthy();
  });

  it("calls onPress when pressed", () => {
    render(<Button title="Click Me" onPress={mockOnPress} />);

    fireEvent.press(screen.getByText("Click Me"));
    expect(mockOnPress).toHaveBeenCalledTimes(1);
  });

  it("renders correctly with small size", () => {
    render(<Button title="Click Me" onPress={mockOnPress} isSmall />);

    expect(screen.getByText("Click Me")).toBeTruthy();
  });

  it("matches snapshot", () => {
    const { toJSON } = render(
      <Button title="Click Me" onPress={mockOnPress} />
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
