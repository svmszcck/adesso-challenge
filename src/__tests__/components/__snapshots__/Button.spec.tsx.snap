// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomButton matches snapshot 1`] = `
<View
  accessibilityLabel="Click Me"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#ffffff",
      "borderRadius": 10,
      "elevation": 8,
      "opacity": 1,
      "paddingHorizontal": 12,
      "paddingVertical": 10,
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#ffffff",
        },
        {
          "fontSize": 16,
          "lineHeight": 24,
        },
        undefined,
        undefined,
        undefined,
        undefined,
        [
          {
            "alignSelf": "center",
            "fontSize": 18,
            "fontWeight": "bold",
          },
          {
            "color": "#009688",
          },
        ],
      ]
    }
  >
    Click Me
  </Text>
</View>
`;
